### PR TITLE
Add `iter_kv_map` lint

### LINKED ISSUE #9376: Lint `map.iter().map(|(_,v)| v)`
### What it does

Warning to iterator operations that are equivalent to `.keys()` and `.values()`.

Similar lint: https://rust-lang.github.io/rust-clippy/stable/index.html#for_kv_map

### Lint Name

iter_kv_map

### Category

style

### Advantage

Simpler and readable than the original code.

### Drawbacks

_No response_

### Example

```rust
let map = HashMap::new();

let keys = map.iter().map(|(key, _)| key).collect::<Vec<_>>();
let values = map.iter().map(|(_, value)| value).collect::<Vec<_>>();
```

Could be written as:

```rust
let map = HashMap::new();

let keys = map.keys().collect::<Vec<_>>();
let values = map.values().collect::<Vec<_>>();
```

### PR DESCRIPTION
**Pull request by DesmondWillowbrook**
fixes #9376

| before | after |
| -------------- | ------------------------- |
| `hmap.iter().map(\|(key, _)\| key)` | `hmap.keys()` |
| `hmap.iter().map(\|(_, v)\| v + 2)` | `hmap.values().map(\|v\| v + 2)` |
| `hmap.into_iter().map(\|(key, _)\| key)` | `hmap.into_keys()` |

Is `MachineApplicable`

changelog: [`iter_kv_map`]: added lint

### CODE REVIEW

File: clippy_lints/src/methods/mod.rs
Comment (daira) [inline clippy_lints/src/methods/mod.rs:3048]:
```suggestion
    /// can be used to express that we only need the keys or that we only need
    /// values.
```

File: clippy_lints/src/methods/iter_kv_map.rs
Comment (daira) [inline clippy_lints/src/methods/iter_kv_map.rs:79]:
```suggestion
/// Returns `true` if the pattern is a `PatWild`, or an ident prefixed with `_`
/// that is not locally used.
```

**Review by daira — APPROVED** [review]
Untested ACK.

**Comment by xFrednet** [comment]
Welcome, Clippy :wave:

There has been a change to `cargo dev update_lints` on master which causes the CI fail. The fix should be as simple as rebasing and running that comment. :upside_down_face:

File: clippy_lints/src/methods/iter_kv_map.rs
Comment (xFrednet) [inline clippy_lints/src/methods/iter_kv_map.rs:31]: You can bind the two patterns directly and check the length like this
```suggestion
        if let PatKind::Tuple([key_pat, val_pat], _) = p.pat.kind;
```

This avoids indexing and is more readable IMO :)

File: clippy_lints/src/methods/iter_kv_map.rs
Comment (xFrednet) [inline clippy_lints/src/methods/iter_kv_map.rs:27]: Small NIT: Currently, we try to move from the `if_chain!` macro to [if let chains](https://rust-lang.github.io/rfcs/2497-if-let-chains.html) if you have the time it would be super to use if let chains instead.

File: clippy_lints/src/methods/iter_kv_map.rs
Comment (xFrednet) [inline clippy_lints/src/methods/iter_kv_map.rs:44]: `snippet_with_context` is intended for lints that also work inside macros. Most lints in Clippy just ignore macros, since It's pretty hard to create a good suggestion and the code might even be outside of the scope that the user can change. Unless you have a specific use case in mind, I would suggest adding a check against macros and then just creating a simple snipped with `snippet_with_applicability`

You can check for macros like this:

```rs
            if expr.span.from_expansion()
```

To include it in the `if_chain` or if-let-chain you can just negate the condition :upside_down_face:

File: clippy_lints/src/methods/iter_kv_map.rs
Comment (xFrednet) [inline clippy_lints/src/methods/iter_kv_map.rs:57]: The message should ideally be a statement. Finding a good message is an art in itself^^

What do think about?
```rs
&format!("usage of `iter` where `{}s()` can be used", replacement_kind)
```

File: clippy_lints/src/methods/iter_kv_map.rs
Comment (xFrednet) [inline clippy_lints/src/methods/iter_kv_map.rs:58]: Optional suggestion:

I prefer shorter help messages like a simple "try", as that makes the console output more readable IMO. But this is mostly a personal preference. You can decide what you prefer and resolve this directly if you want :)

**Review by xFrednet — COMMENTED** [review]
Overall looks good to me and I really like this lint! I made some style related comments, which should be simple to fix. You're welcome to reach out if you need any help with them :)

File: tests/ui/iter_kv_map.rs
Comment (xFrednet) [reply tests/ui/iter_kv_map.rs:29]: Could you add a test, where the value is named but never used like this:

```rs
// Lint
map.iter().map(|(key, _value)| key * 9).count();
map.iter().map(|(_key, value)| value * 17).count();
```

File: clippy_lints/src/methods/iter_kv_map.rs
Comment (DesmondWillowbrook) [reply clippy_lints/src/methods/iter_kv_map.rs:57]: How about `&format!("iterating on a map's {}s", replacement_kind)`?

Comment (xFrednet) [reply clippy_lints/src/methods/iter_kv_map.rs:57]: would be fine with me :upside_down_face:

File: tests/ui/iter_kv_map.rs
Comment (DesmondWillowbrook) [reply tests/ui/iter_kv_map.rs:24]: Yes, I undestand that - which is why I was surprised that this _only_ checks method calls instead of the desugared syntax. It wasn't really my intention to test against this behaviour; I'm removing it now.

File: clippy_lints/src/methods/mod.rs
Comment (DesmondWillowbrook) [reply clippy_lints/src/methods/mod.rs:3056]: Ah, not really. I just copied stuff from the `for_kv_map` lint, which uses `ignore`.

File: tests/ui/iter_kv_map.rs
Comment (xFrednet) [reply tests/ui/iter_kv_map.rs]: The fact that its `Expr::MethodCall` is basically the desugared syntax. There is a PR in rustc to have the self argument as an extra member in the enum. But I'm not sure what the status if that one is :)

File: clippy_lints/src/methods/mod.rs
Comment (xFrednet) [reply clippy_lints/src/methods/mod.rs:3066]: You've resolved this comment. Is there a reason why we would want to keep this lint in `nursery`?

File: clippy_lints/src/methods/mod.rs
Comment (xFrednet) [reply clippy_lints/src/methods/mod.rs:3057]: In that case, could you remove the `ignore` so we test the code examples in our CI?

File: tests/ui/iter_kv_map.rs
Comment (xFrednet) [inline tests/ui/iter_kv_map.rs:26]:
```suggestion
    let _ = map.clone().into_iter().map(|(_, val)| val).collect::<Vec<_>>();
    let _ = map.clone().into_iter().map(|(_, val)| val + 2).collect::<Vec<_>>();
```

**Review by xFrednet — COMMENTED** [review]
You seem to have resolved some comments, even if they haven't been addressed or a comment why you think that they don't apply. I've now unresolved them. If you have questions or are unsure what I mean by them, please say so. Please tell me, if I misunderstood something.

This review sadly took some time, I'll try to reply faster next time.

Another note, Clippy uses rusts no merge commit policy. Merge commits are fine for now, but will need to be dropped or squashed before the merge. That's a bridge we can cross when we come to it :upside_down_face:

File: clippy_lints/src/methods/mod.rs
Comment (xFrednet) [reply clippy_lints/src/methods/mod.rs:3057]: In that case, could you remove the `ignore` so we test the code examples in our CI?

File: clippy_lints/src/methods/mod.rs
Comment (xFrednet) [reply clippy_lints/src/methods/mod.rs:3066]: You've resolved this comment. Is there a reason why we would want to keep this lint in nursery?

File: tests/ui/iter_kv_map.rs
Comment (DesmondWillowbrook) [reply tests/ui/iter_kv_map.rs:39]: Ah, sorry. I forgot to do the second part. Done now!

File: clippy_lints/src/methods/mod.rs
Comment (DesmondWillowbrook) [reply clippy_lints/src/methods/mod.rs]: Hmm, I can swear I remember doing this...
Done now! 🙃

File: clippy_lints/src/methods/mod.rs
Comment (DesmondWillowbrook) [reply clippy_lints/src/methods/mod.rs]: Done.

File: clippy_lints/src/methods/mod.rs
Comment (DesmondWillowbrook) [reply clippy_lints/src/methods/mod.rs]: Fixed now.

I would imagine that these might be optimized by the compiler, so the only way to know for sure is to benchmark it.

**Comment by DesmondWillowbrook** [comment]
> You seem to have resolved some comments, even if they haven't been addressed
Really sorry - I don't know what was going through my mind when I did that! Rest assured, I'll try to be more attentive moving forwards.

> This review sadly took some time, I'll try to reply faster next time.
Please don't say that - I'm used to reviews taking a week, so this was super quick for me.

> Clippy uses rusts no merge commit policy. 
Ah, acknowledged. Should I squash/rebase right now?

**Comment by xFrednet** [comment]
> Really sorry - I don't know what was going through my mind when I did that! Rest assured, I'll try to be more attentive moving forwards.

No worries, I was just confused. This happens every once in a while :)

> > This review sadly took some time, I'll try to reply faster next time.
> 
> Please don't say that - I'm used to reviews taking a week, so this was super quick for me.

Thank you for the encouragement! I try to review every PR within 5 days. But life got the better of me, so now I'm just catching up on everything :). I'll give this a full review over the week :)

> Ah, acknowledged. Should I squash/rebase right now?

You can do that, like you did, if I see it correctly. For larger PR it's better to wait until the review is done since GitHub is a bit iffy when it comes to linking old review comments to rebased code. But I remember them in this case, and they should all be addressed.

File: clippy_lints/src/methods/iter_kv_map.rs
Comment (xFrednet) [inline clippy_lints/src/methods/iter_kv_map.rs:68]: I would adjust this lint emission to have the same message as the one above :upside_down_face: 
```suggestion
                        &format!("iterating on a map's {}s", replacement_kind),
                        "try",
```

File: tests/ui/iter_kv_map.rs
Comment (xFrednet) [inline tests/ui/iter_kv_map.rs:9]: Could you add a test with a local closure or external function like this:

```rs
fn magic_map(_: u32, val: u32) -> u32 {
	val
}

// ...

fn main() {
	let some_closure = |(key: u32, _val: u32)| key;

	// For boths maps:

	// Don't lint external funciton
	map.iter().map(magic_map).collect::<Vec<_>>();
	// Don't lint external closure
	map.iter().map(some_closure).collect::<Vec<_>>();
}
```

Here it's fine to only check for the `value` or `key` since this shouldn't be linted. I believe the current implementation already deals with this correctly.

---

One more test I could think of was this:

```rs
map.iter().map(|(_, val)| {val}).collect::<Vec<_>>();
```

This will most likely suggest `values().map(|val| {val})` which is totally fine. That will be caught by the identity lint, if the user doesn't realize that they can remove the second map :upside_down_face:

**Review by xFrednet — COMMENTED** [review]
Alright, this version already looks good to me. I had a few more suggestions for test, and then we can get this merged.

Could you also squash your commits? It's fine to have more than one if they're named properly, or just to have one review commit. :upside_down_face: 

Small side note: I ran this on 26 Crates with our lintcheck, and it identified 5 cases, which were all correct. So it looks like time implementation is good to go :tada:

**Comment by DesmondWillowbrook** [comment]
> Could you also squash your commits? 

Squashed.

>  I ran this on 26 Crates with our lintcheck, and it identified 5 cases

Huh, and here I thought this was too niche a lint. Happy to hear that!

**Comment by DesmondWillowbrook** [comment]
Oof, a pre-commit hook or maybe `cargo uitest` or something is automatically formatting the identity block whenever I try to commit it. It's messing up the tests!

**Comment by xFrednet** [comment]
> Oof, a pre-commit hook is correcting `|(_, val)| {val}` to `|(_, val)| val`. Even worse, adding `#[rustfmt::skip]` doesn't help because then it disables the rustfix tests. Trying to add the attribute to the statement shows me that that feature is still in nightly.
> 
> Would it be fine to just skip this test? @xFrednet

Alright, then it's fine if we skip that test. Could you fix the CI? Then we can merge it :)

**Review by xFrednet — APPROVED** [review]

**Comment by xFrednet** [comment]
Looks good to me, thank you for this implementation. The lint will be a great addition to Clippy. I hope you had fun while working on Clippy :upside_down_face: 

@bors r+

**Pull request merged**
